uxarray/_mpas.py

```diff
@@ -0,0 +1,356 @@
+    for vert_idx in range(maxEdges):
+        # mask for non-padded values
+        mask = vert_idx < nEdgesOnCell
+        # replace remaining padding or zeros with fill_value
+        verticesOnCell[np.logical_not(mask), vert_idx] = fill_val
+
+    return verticesOnCell
+
+
+def _replace_zeros(grid_var):
+    """Replaces all instances of a zero (invalid/missing MPAS value) with a
+    fill value.
+
+    Parameters
+    ----------
+    grid_var : numpy.ndarray
+        Grid variable that may contain zeros that need to be replaced
+
+    Returns
+    -------
+    grid_var : numpy.ndarray
+        Grid variable with zero replaced by fill values, done in-place
+    """
+
+    # one dimensional view of grid variable
+    grid_var_flat = grid_var.ravel()
+
+    # replace all zeros with a fill value
+    grid_var_flat[grid_var_flat == 0] = fill_val
+
+    return grid_var
+
+
+def _to_zero_index(grid_var):
+    """Given an input using that is one-indexed, subtracts one from all non-
+    fill value entries to convert to zero-indexed.
+
+    Parameters
+    ----------
+    grid_var : numpy.ndarray
+        Grid variable that is one-indexed
+
+    Returns
+    -------
+    grid_var : numpy.ndarray
+        Grid variable that is converted to zero-indexed, done in-place
+    """
+    # one dimensional view of grid variable
+    grid_var_flat = grid_var.ravel()
+
+    # convert non-fill values to zero-indexed
+    grid_var_flat[grid_var_flat != fill_val] -= 1
+
+    return grid_var
+
+
+def _read_mpas(ext_ds, use_dual=False):
+    """Function to read in a MPAS Grid dataset and encode either the Primal or
+    Dual mesh in the UGRID conventions.
+
+    Parameters
+    ----------
+    ext_ds : xarray.Dataset, required
+        MPAS datafile of interest
+    use_dual : bool, optional
+        Flag to select whether to encode the Dual-Mesh. Defaults to False
+
+    Returns
+    -------
+    ds : xarray.Dataset
+        ugrid aware :class:`xarray.Dataset`
+    """
+
+    # empty dataset that will contain our encoded MPAS mesh
+    ds = xr.Dataset()
+
+    try:
+        # convert dual-mesh to UGRID
+        if use_dual:
+            _dual_to_ugrid(ext_ds, ds)
+        # convert primal-mesh to UGRID
+        else:
+            _primal_to_ugrid(ext_ds, ds)
+
+        ds["Mesh2"] = xr.DataArray(
+            attrs={
+                "cf_role": "mesh_topology",
+                "long_name": "Topology data of unstructured mesh",
+                "topology_dimension": 2,
+                "node_coordinates": "Mesh2_node_x Mesh2_node_y",
+                "node_dimension": "nMesh2_node",
+                "face_node_connectivity": "Mesh2_face_nodes",
+                "face_dimension": "nMesh2_face"
+            })
+    except:
+        raise Exception(
+            "Variables not in recognized MPAS form. Please refer to",
+            "https://mpas-dev.github.io/files/documents/MPAS-MeshSpec.pdf",
+            "for more information on MPAS Grid file formatting")
```

Comment:
I'm a bit concerned that this exception is broad enough, and there's enough that could go wrong in the try block, that this error message might not appropriately reflect what the root cause of an issue may be. 

I ran into this situation when I was running the tests while messing with some of the suggestions later in this review. I had made a typo in a variable name and the tests all triggered this exception message.

Would it be possible to wrap only the parts of `_dual_to_ugrid` and `_primal_to_ugrid` that are extracting the MPAS attributes in a try/catch instead of the entire read-in process?

Reply:
Oh, I missed this but I agree with @anissa111 that the `try`/`except` is covering too much code.  In addition, it is best to avoid raising a generic `Exception` because as @anissa111 says, it catches *everything* including typos in the code or even someone trying to abort the code with `^c` (a `KeyboardInterrupt`).  If existing exceptions like `ValueError` or `OSError` aren't sufficient, you should strongly consider implementing your own exception.

Reply:
Good points, I didn't give this part enough attention (based it off what was already implemented in `_scrip.py`).

I'll work on setting up some more robust catches.

Reply:
It might be better not to have this `try`/`except` at all and to focus on individual errors within each function that are more specific to the problems actually triggering the error.  As a user of `uxarray`, I wouldn't find this blanket exception very useful because it would replace a different exception that could point me toward the problem in more detail.

Reply:
I've removed the `try` / `except` entirely. 

Once we have more robust grid format checks, we should be able to handle errors and exceptions outside of these I/O modules. (see #238)

Example:
* User inputs a subset of an MPAS grid that excludes necessary variables to convert to UGRID
* We should be able to catch this before jumping into `_read_mpas()` and raise an appropriate exception

Reply:
That sounds like a much cleaner plan  Thanks!